internal/provisioner/objects/service/service.go

```diff
@@ -108,78 +108,62 @@ var (
 	}
 )
 
-// EnsureContourService ensures that a Contour Service exists for the given contour.
-func EnsureContourService(ctx context.Context, cli client.Client, contour *model.Contour) error {
-	desired := DesiredContourService(contour)
-	current, err := currentContourService(ctx, cli, contour)
+type desiredService func(*model.Contour) *corev1.Service
+type currentService func(context.Context, client.Client, *model.Contour) (*corev1.Service, error)
+type serviceUpdater func(context.Context, client.Client, *model.Contour, *corev1.Service, *corev1.Service) error
+
+func ensureServiceHelper(ctx context.Context, cli client.Client, contour *model.Contour, desiredSvc desiredService, currentSvc currentService, updater serviceUpdater) error {
+	desired := desiredSvc(contour)
+	current, err := currentSvc(ctx, cli, contour)
 	if err != nil {
 		if errors.IsNotFound(err) {
 			return createService(ctx, cli, desired)
 		}
 		return fmt.Errorf("failed to get service %s/%s: %w", desired.Namespace, desired.Name, err)
 	}
-	if err := updateContourServiceIfNeeded(ctx, cli, contour, current, desired); err != nil {
+	if err := updater(ctx, cli, contour, current, desired); err != nil {
 		return fmt.Errorf("failed to update service %s/%s: %w", desired.Namespace, desired.Name, err)
 	}
 	return nil
 }
 
+// EnsureContourService ensures that a Contour Service exists for the given contour.
+func EnsureContourService(ctx context.Context, cli client.Client, contour *model.Contour) error {
+	return ensureServiceHelper(ctx, cli, contour, DesiredContourService, currentContourService, updateContourServiceIfNeeded)
+}
+
 // EnsureEnvoyService ensures that an Envoy Service exists for the given contour.
 func EnsureEnvoyService(ctx context.Context, cli client.Client, contour *model.Contour) error {
-	desired := DesiredEnvoyService(contour)
-	current, err := currentEnvoyService(ctx, cli, contour)
+	return ensureServiceHelper(ctx, cli, contour, DesiredEnvoyService, currentEnvoyService, updateEnvoyServiceIfNeeded)
+}
+
+func ensureServiceDeletedHelper(ctx context.Context, cli client.Client, contour *model.Contour, currentSvc currentService) error {
```

Comment:
Let's call this just `ensureServiceDeleted`. Also, instead of taking a `currentService` func, let's just take a namespaced name